geo/src/algorithm/mean_center.rs

```diff
@@ -0,0 +1,694 @@
+                MultiPoint::new(centroids).centroid()
+            }
+
+            fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output {
+                let centroids = self
+                    .iter()
+                    .map(|p| p.centroid())
+                    .filter(|c| c.is_some())
+                    .map(|c| c.unwrap())
+                    .collect::<Vec<_>>();
+
+                let n = centroids.len();
+
+                if n == 0 {
+                    return None;
+                }
+
+                // Create multipoint from the centroids and calculate the mean center
+                MultiPoint::new(centroids).weighted_mean_center(weights)
+            }
+        }
+    };
+}
+
+// Use the macro for MultiPolygon and MultiLineString
+impl_mean_center_multi!(MultiPolygon);
+impl_mean_center_multi!(MultiLineString);
+
+impl MeanCenter for Geometry {
+    type Output = Option<Point>;
+    fn mean_center(&self) -> Self::Output {
+        match self {
+            Geometry::Point(p) => p.mean_center().into(),
+            Geometry::Line(l) => l.mean_center().into(),
+            Geometry::LineString(ls) => ls.mean_center(),
+            Geometry::Polygon(p) => p.mean_center(),
+            Geometry::MultiPoint(mp) => mp.mean_center(),
+            Geometry::MultiLineString(mls) => mls.mean_center(),
+            Geometry::MultiPolygon(mp) => mp.mean_center(),
+            Geometry::GeometryCollection(gc) => gc.mean_center(),
+            Geometry::Rect(r) => r.mean_center().into(),
+            Geometry::Triangle(t) => t.mean_center().into(),
+        }
+    }
+
+    fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output {
+        match self {
+            Geometry::Point(p) => p.weighted_mean_center(weights).into(),
+            Geometry::Line(l) => l.weighted_mean_center(weights).into(),
+            Geometry::LineString(ls) => ls.weighted_mean_center(weights),
+            Geometry::Polygon(p) => p.weighted_mean_center(weights),
+            Geometry::MultiPoint(mp) => mp.weighted_mean_center(weights),
+            Geometry::MultiLineString(mls) => mls.weighted_mean_center(weights),
+            Geometry::MultiPolygon(mp) => mp.weighted_mean_center(weights),
+            Geometry::GeometryCollection(gc) => gc.weighted_mean_center(weights),
+            Geometry::Rect(r) => r.weighted_mean_center(weights).into(),
+            Geometry::Triangle(t) => t.weighted_mean_center(weights).into(),
+        }
+    }
+}
+
+impl MeanCenter for GeometryCollection {
+    type Output = Option<Point>;
+    fn mean_center(&self) -> Self::Output {
+        // take the centroid of each geometry
+        // the mean center is calculated from the centroids
+        // any geometry that does not have a centroid is filtered out
+        // this should be rare
+        let centroids = self
+            .iter()
+            .map(|g| g.centroid())
+            .filter(|c| c.is_some())
```

Comment:
You can replace the `map/filter/map(unwrap)` with `flat_map(|g| g.centroid())`

(this occurs in several places in this PR)